File: Src/HashTable/hash.zig
const std = @import("std");

pub fn HashTable(comptime Key: type, comptime Value: type) type {
    return struct {
        const Self = @This();

        allocator: std.mem.Allocator,
        size: usize,
        buckets: std.ArrayList(std.ArrayList(Entry)),

        const Entry = struct {
            key: Key,
            value: Value,
        };

        pub fn init(allocator: std.mem.Allocator, size: usize) !Self {
            if (size == 0) return error.InvalidSize;

            var buckets = std.ArrayList(std.ArrayList(Entry)).init(allocator);
            errdefer buckets.deinit();

            for (0..size) |_| {
                try buckets.append(std.ArrayList(Entry).init(allocator));
            }

            return Self{
                .allocator = allocator,
                .size = size,
                .buckets = buckets,
            };
        }

        pub fn deinit(self: *Self) void {
            for (self.buckets.items) |*bucket| {
                bucket.deinit();
            }
            self.buckets.deinit();
        }

        fn getIndex(self: Self, key: Key) usize {
            const hash_u64 = std.hash.autoHash(key);
            return @as(usize, @intCast(hash_u64 % self.size));
        }

        pub fn remove(self: *Self, key: Key) ?Value {
            const index = self.getIndex(key);
            const bucket = &self.buckets.items[index];

            for (bucket.items, 0..) |entry, i| {
                if (entry.key == key) {
                    const removed_value = entry.value;
                    _ = bucket.swapRemove(i);
                    return removed_value;
                }
            }

            return null;
        }

        pub fn search(self: Self, key: Key) ?Value {
            const index = self.getIndex(key);
            const bucket = &self.buckets.items[index];

            for (bucket.items) |entry| {
                if (entry.key == key) {
                    return entry.value;
                }
            }

            return null;
        }
        pub fn insert(self: *Self, key: Key, value: Value) !void {
            const index = self.getIndex(key);
            const bucket = &self.buckets.items[index];

            for (bucket.items) |*entry| {
                if (entry.key == key) {
                    entry.value = value;
                    return;
                }
            }

            try bucket.append(.{ .key = key, .value = value });
        }

        pub fn update(self: *Self, key: Key, value: Value) !void {
            const index = self.getIndex(key);
            const bucket = &self.buckets.items[index];

            for (bucket.items) |*entry| {
                if (entry.key == key) {
                    entry.value = value;
                    return;
                }
            }

            return error.KeyNotFound;
        }
    };
}

File: Src/HashTable/hash.py
class HashTable:
    def __init__(self, size: int, hash_function: object = None):
        """
        Constructor de la clase HashTable usando encadenamiento.
        
        :param size: Tamaño de la tabla hash.
        :param hash_function: Función hash personalizada; si no se proporciona, se usa hash(key) % size.
        """
        self.size = size
        self.table = [[] for _ in range(size)]  # Cada casilla es una lista vacía.
        self.hash_function = hash_function if hash_function is not None else (lambda key: hash(key) % size)

    def insert(self, key, value):
        """
        Inserta el par (clave, valor) en la tabla.
        Si la clave ya existe, se actualiza el valor.
        """
        index = self.hash_function(key)
        for i, (k, v) in enumerate(self.table[index]):
            if k == key:
                self.table[index][i] = (key, value)
                return
        self.table[index].append((key, value))

    def search(self, key):
        """
        Busca la clave y retorna su valor asociado; si no existe, retorna None.
        """
        index = self.hash_function(key)
        for k, v in self.table[index]:
            if k == key:
                return v
        return None

    def remove(self, key):
        """
        Elimina el par (clave, valor) de la tabla.
        Retorna el valor eliminado o None si la clave no existe.
        """
        index = self.hash_function(key)
        for i, (k, v) in enumerate(self.table[index]):
            if k == key:
                del self.table[index][i]
                return v
        return None

    def update(self, key, value):
        """
        Actualiza el valor asociado a la clave.
        Lanza KeyError si la clave no se encuentra.
        """
        index = self.hash_function(key)
        for i, (k, _) in enumerate(self.table[index]):
            if k == key:
                self.table[index][i] = (key, value)
                return
        raise KeyError("Clave no encontrada")

    def __str__(self):
        """
        Retorna una representación en cadena de la tabla hash.
        """
        result = []
        for i, bucket in enumerate(self.table):
            result.append(f"{i}: {bucket}")
        return "\n".join(result)

